syne_tune/optimizer/schedulers/multiobjective/nsga2_searcher.py

```diff
@@ -36,26 +36,35 @@
     print(try_import_moo_message())
 
 
-class MultiObjectiveMixedVariableProblem(Problem):
-    def __init__(self, n_obj: int, config_space: Dict[str, Any], **kwargs):
-        vars = {}
-
-        for hp_name, hp in config_space.items():
-            if isinstance(hp, Domain):
-                if isinstance(hp, Categorical):
-                    vars[hp_name] = PyMOOChoice(options=hp.categories)
-                elif isinstance(hp, Integer):
-                    vars[hp_name] = PyMOOInteger(bounds=(hp.lower, hp.upper - 1))
-                elif isinstance(hp, FiniteRange):
-                    vars[hp_name] = PyMOOInteger(bounds=(0, hp.size - 1))
-                elif isinstance(hp, Float):
-                    vars[hp_name] = PyMOOReal(bounds=(hp.lower, hp.upper))
-                else:
-                    raise Exception(
-                        f"Type {type(hp)} of hyperparameter {hp_name} is not supported!"
-                    )
-
-        super().__init__(vars=vars, n_obj=n_obj, n_ieq_constr=0, **kwargs)
+def _create_multiobjective_problem(config_space: Dict[str, Any], n_obj: int, **kwargs):
+    # This needs to be an inner class, since ``Problem`` can only be imported
+    # with ``moo`` dependencies. We want this module to be importable even if
+    # ``moo`` dependencies are not present: only creating a ``NSGA2Searcher``
+    # object should fail in this case.
+    class _MultiObjectiveMixedVariableProblem(Problem):
+        def __init__(self, n_obj: int, config_space: Dict[str, Any], **kwargs):
+            vars = {}
+
+            for hp_name, hp in config_space.items():
+                if isinstance(hp, Domain):
+                    if isinstance(hp, Categorical):
+                        vars[hp_name] = PyMOOChoice(options=hp.categories)
+                    elif isinstance(hp, Integer):
+                        vars[hp_name] = PyMOOInteger(bounds=(hp.lower, hp.upper - 1))
+                    elif isinstance(hp, FiniteRange):
+                        vars[hp_name] = PyMOOInteger(bounds=(0, hp.size - 1))
+                    elif isinstance(hp, Float):
+                        vars[hp_name] = PyMOOReal(bounds=(hp.lower, hp.upper))
+                    else:
+                        raise Exception(
+                            f"Type {type(hp)} of hyperparameter {hp_name} is not supported!"
```

Comment:
The current change of nesting this in a class makes sense to fix the import issue.

But all these `isinstance()` checks added in https://github.com/awslabs/syne-tune/pull/660 feel somehow brittle to me, swapping in all these PyMOO datatypes here. I wonder if there's a cleaner way to use MOO, or if using MOO necessarily introduces this complexity. 

Why does PyMOO even create those special pymoo datatypes?

Reply:
Well, fair enough, we also create our own datatypes!

Reply:
I think this is fine. As long as we remember to keep this stuff in functions, so it is not imported when not really needed. And the new integ tests will make sure this does not happen